pynicotine/gtkgui/dialogs/downloadfolder.py

```diff
@@ -0,0 +1,253 @@
+from pynicotine.gtkgui.widgets import ui
+from pynicotine.gtkgui.widgets.dialogs import Dialog
+from pynicotine.gtkgui.widgets.treeview import TreeView
+from pynicotine.gtkgui.widgets.infobar import InfoBar
+from pynicotine.gtkgui.widgets.filechooser import FolderChooser
+from pynicotine import slskmessages
+from pynicotine.utils import human_size
+
+
+class DownloadFolder(Dialog):
+
+    def __init__(self, application):
+
+        (
+            self.container,
+            self.destination_entry,
+            self.info_bar,
+            self.list_container,
+            self.virtual_path
+        ) = ui.load(scope=self, path="dialogs/downloadfolder.ui")
+
+        super().__init__(
+            parent=application.window,
+            modal=False,
+            content_box=self.container,
+            title="Folder File Download",
+            width=800,
+            height=600,
+            close_callback=self.on_close,
+            close_destroy=False
+        )
+
+        self.info_bar = InfoBar(self.info_bar)
+        self.application = application
+        self.folder_files = None
+
+        self.list_view = TreeView(
+            application.window, parent=self.list_container, multi_select=True, activate_row_callback=None,
+            columns={
+                "filename": {
+                    "column_type": "text",
+                    "title": "File Name",
+                    "width": 1,
+                    "expand_column": True
+                },
+                "size": {
+                    "column_type": "text",
+                    "title": "Size",
+                    "width": 100,
+                    "expand_column": False
+                },
+                "download": {
+                    "column_type": "toggle",
+                    "title": "Download",
+                    "width": 0,
+                    "toggle_callback": self.on_download_option
+                }
+            }
+        )
+
+        self.info_bar.show_message("Loading", message_type=Gtk.MessageType.INFO)
+
+        for event_name, callback in (
+            ("folder-contents-response", self._folder_contents_response),
+            ("peer-connection-error", self._peer_connection_error)
+        ):
+            events.connect(event_name, callback)
+
+    def _peer_connection_error(self, user, msgs=None, is_offline=False):
+
+        if msgs is None:
+            return
+
+        for i in msgs:
+            if i.__class__ is slskmessages.FolderContentsRequest:
+                self.toggle_enabled_view(False)
+                self.info_bar.show_message(user + " " + "User logged off" if is_offline else "Connection timeout",
+                                           message_type=Gtk.MessageType.ERROR)
+
+    def toggle_enabled_view(self, valid_files):
+
+        self.container.set_sensitive(valid_files)
+        if (valid_files):
+            self.info_bar.set_visible(False)
+
+    # Pass the Download Folder Data
```

Comment:
These function definition header comments are not required, your code is self-explanatory.